google/cloud/bigtable/mutations_batcher.py

```diff
@@ -46,61 +177,322 @@ class MutationsBatcher:
+        # keep track of the first and last _exception_list_limit exceptions
+        self._exception_list_limit: int = 10
+        self._oldest_exceptions: list[Exception] = []
+        self._newest_exceptions: deque[Exception] = deque(
+            maxlen=self._exception_list_limit
+        )
+        # clean up on program exit
+        atexit.register(self._on_exit)
+
+    def _start_flush_timer(self, interval: float | None) -> asyncio.Future[None]:
+        """
+        Set up a background task to flush the batcher every interval seconds
+
+        If interval is None, an empty future is returned
 
-    async def append(self, row_key: str | bytes, mutation: Mutation | list[Mutation]):
+        Args:
+          - flush_interval: Automatically flush every flush_interval seconds.
+              If None, no time-based flushing is performed.
+        Returns:
+            - asyncio.Future that represents the background task
         """
-        Add a new mutation to the internal queue
+        if interval is None or self.closed:
+            empty_future: asyncio.Future[None] = asyncio.Future()
+            empty_future.set_result(None)
+            return empty_future
+
+        async def timer_routine(self, interval: float):
+            """
+            Triggers new flush tasks every `interval` seconds
+            """
+            while not self.closed:
+                await asyncio.sleep(interval)
+                # add new flush task to list
+                if not self.closed and self._staged_entries:
+                    self._schedule_flush()
+
+        timer_task = asyncio.create_task(timer_routine(self, interval))
+        return timer_task
+
+    async def append(self, mutation_entry: RowMutationEntry):
         """
-        raise NotImplementedError
+        Add a new set of mutations to the internal queue
 
-    async def append_conditional(
-        self,
-        predicate_filter: RowFilter,
-        row_key: str | bytes,
-        if_true_mutations: Mutation | list[Mutation] | None = None,
-        if_false_mutations: Mutation | list[Mutation] | None = None,
-    ):
+        Args:
+          - mutation_entry: new entry to add to flush queue
+        Raises:
+          - RuntimeError if batcher is closed
+          - ValueError if an invalid mutation type is added
+        """
+        if self.closed:
+            raise RuntimeError("Cannot append to closed MutationsBatcher")
+        if isinstance(mutation_entry, Mutation):  # type: ignore
+            raise ValueError(
+                f"invalid mutation type: {type(mutation_entry).__name__}. Only RowMutationEntry objects are supported by batcher"
+            )
+        self._staged_entries.append(mutation_entry)
+        # start a new flush task if limits exceeded
+        self._staged_count += len(mutation_entry.mutations)
+        self._staged_bytes += mutation_entry.size()
+        if (
+            self._staged_count >= self._flush_limit_count
```

Comment:
I think we should check if staged_count + mutation_metry.mutations > self.flush_limit_count first before adding it to the staged_entries. Otherwise we could exceed the flush_limit_count?